Feature/FA-11-Most-Selling-Products

EslamMagdy12 opened the pull request:
This pull request introduces a new feature for handling the "Most Selling Products" in the e-commerce flower app. It includes changes to the data layer, domain layer, and presentation layer to support fetching and displaying the best-selling products.

### Changes Included:

#### 🔹 Data Layer:
- Added API endpoints in `BestSellerRetrofitClient` for:
  - Retrieving best seller products .
- Implemented **Best Seller Remote Data Source** to manage best-seller-products-related API calls.
- Created response model for:
  - `best_seller_products`
- Developed the **Best Seller Repository Implementation** to handle data fetching and processing.

#### 🔹 Domain Layer:
- Created new **use cases** for handling:
  - Best seller products requests.
- Integrated **repository contracts** to ensure structured data handling.

#### 🔹 Presentation Layer:
- Developed dedicated screens for:
  - **Best Seller Products**
- Developed widget for `ProductItem`.

## 📌 Pull Request Checklist

### ✅ General
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] The PR title follows the required format: `type/FA-<issue-number>-TaskName`.
- [x] The branch name follows the required format: `feature/meaningful-name`, `fix/issue-description`, etc.

### 📂 Code Quality
- [x] All filenames are in `snake_case`.
- [x] All functions and variable names follow Dart naming conventions.
- [x] All new code is formatted with `dart format .`

### 🛠 CI/CD Validation
- [x] Linting passes with `dart analyze`.
- [x] The branch name is validated against the allowed patterns.
- [x] The PR title is validated against the required format.

### 🚀 Testing
- [ ] Unit tests are included for all new features and bug fixes.
- [ ] `flutter test --coverage` has been run and the results reviewed.
- [ ] The project builds successfully without errors.

### 🔗 Related JIRA Ticket(s):
- [x] JIRA Issue: `FA-11`

### 📸 Screenshots (if applicable):
<div align="center">

|         Best Selling Products          |
|:---------------------:|
| <img src="https://github.com/user-attachments/assets/3f520805-f024-4fe9-9cd5-f5b59cce4abd" width="300" alt="Best Selling Products"> | 

</div>

EslamMagdy12 committed:
feat(best_seller): Implement best selling products feature.

- Added data layer with models, API client, and remote data source.
- Created repository implementation for handling best seller data.

EslamMagdy12 committed:
feat(best_seller_products): Implement domain layer for best selling products

- Add BestSellerProductEntity for product data.
- Create BestSellerRepo for data access abstraction.
- Define GetBestSellersUsecase for retrieving best sellers.

EslamMagdy12 committed:
feat(best_sellers): Implement best seller products feature presentation layer.

- Added `BestSellerProductEntity` to represent best-selling product data.
- Created `ProductItem` widget to display individual product information.
- Implemented `BestSellers` screen to display a list of best-selling products using a `GridView`.
- Added `BestSellerCubit` and `BestSellerState` for state management and handling data fetching.
- Integrated `GetBestSellersUsecase` to fetch best-selling products.

EslamMagdy12 committed:
feat(core, most_selling_products): Integrate best selling products feature

- Updated Android build configuration to specify NDK version.
- Added route for best seller screen in the app.
- Set initial route to best seller screen for testing.
- Registered dependencies for best seller feature using dependency injection.

EslamMagdy12 committed:
refactor(best_selling-products): reformat and analyze codes using dart format and dart analyze.

ali72-20 commented on lib/features/most_selling_products/data/data_source/contract/best_seller_remote_data_source.dart:
```diff
@@ -0,0 +1,5 @@
+import '../../model/best_seller_products.dart';
+
+abstract class BestSellerRemoteDataSource {
```
interface more better than abstract class if you don't have any attributes and it only methods

ali72-20 commented on lib/features/most_selling_products/data/data_source/remote/best_seller_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,17 @@
+
+@Injectable(as: BestSellerRemoteDataSource)
+class BestSellerRemoteDataSourceImpl implements BestSellerRemoteDataSource {
+  final BestSellerRetrofitClient bestSellerRetrofitClient;
```
make it private that the only class can acess retrofit client is data source so private more better

ali72-20 commented on lib/features/most_selling_products/domain/entity/best_seller_product_entity.dart:
```diff
@@ -0,0 +1,15 @@
+class BestSellerProductEntity {
```
if the response model has nullable attribute you need to make this attribute also nullable

ali72-20 commented on lib/features/most_selling_products/domain/usecase/get_best_sellers_usecase.dart:
```diff
@@ -0,0 +1,16 @@
+
+@injectable
+class GetBestSellersUsecase {
+  final BestSellerRepo bestSellerRepo;
```
make it private more better that this use case the only can access best seller repo in this scope

ali72-20 commented on lib/features/most_selling_products/presentation/view/widgets/product_item.dart:
```diff
@@ -0,0 +1,95 @@
+                  Row(
+                    children: [
+                      Text(
+                        'EGP ${productEntity.priceAfterDiscount}',
```
EGP is hard coded in arabic should be جنيه مصزي so use local here

ali72-20 commented on lib/features/most_selling_products/presentation/view_model/best_seller_cubit.dart:
```diff
@@ -0,0 +1,39 @@
+
+@injectable
+class BestSellerCubit extends Cubit<BestSellerState> {
+  final GetBestSellersUsecase getBestSellersUsecase;
```
make private more better so no any one can access this use case best seller cubit only class can do this

ali72-20 reviewed (approved):
small changes make it more better 
great work keep going

EslamMagdy12 committed:
refactor(best_seller): refactor best-selling products and handle empty state
- Replaced public variables with private ones in:
    - BestSellerRemoteDataSourceImpl
    - GetBestSellersUsecase
    - BestSellerCubit
  - This enhances encapsulation and prevents unintended external access.
- Shows a "No products found" message when the list is empty.
- Displays an error dialog if fetching products fails.
- Replaced hard coded string with its localization.

EslamMagdy12 committed:
refactor(data_source): Change BestSellerRemoteDataSource from abstract class to abstract interface class

EslamMagdy12 committed:
Merge remote-tracking branch 'origin/developement' into feature/fa-11-most-selling-products

# Conflicts:
#	assets/translations/ar.json
#	assets/translations/en.json
#	lib/core/utils/di/di.config.dart
#	lib/core/utils/l10n/locale_keys.g.dart
#	lib/core/utils/routes/routes.dart

EslamMagdy12 committed:
refactor(most_selling_products): removed hard coded strings.

ahmedelazab1220 commented on lib/features/most_selling_products/data/api/best_seller_retrofit_client.dart:
```diff
@@ -0,0 +1,18 @@
+
+part 'best_seller_retrofit_client.g.dart';
+
+@singleton
```
make injectable is more better

EslamMagdy12 replied:
This is APIClient it's better to be annotated with singleton I just want only the same exact instance when calling it from datasource, if I made it injectable it will create a new instance everytime I do api call

ahmedelazab1220 commented on lib/features/most_selling_products/data/model/best_seller_products.dart:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+
+@JsonSerializable()
+class BestSellerProduct {
+  @JsonKey(name: "_id")
```
make in seprate file is better instead of make two model in one class

ahmedelazab1220 commented on lib/features/most_selling_products/data/repo_impl/best_seller_repo_impl.dart:
```diff
@@ -0,0 +1,22 @@
+
+@Injectable(as: BestSellerRepo)
+class BestSellerRepoImpl implements BestSellerRepo {
+  final ApiManager _apiManager = ApiManager();
```
inject by constructor is better

ahmedelazab1220 commented on lib/features/most_selling_products/presentation/view/screens/best_sellers.dart:
```diff
@@ -0,0 +1,83 @@
+import '../../../../../core/utils/l10n/locale_keys.g.dart';
+import '../../view_model/best_seller_state.dart';
+
+class BestSellers extends StatelessWidget {
```
name must be BestSellersScreen not BestSellers

ahmedelazab1220 commented on lib/features/most_selling_products/presentation/view/screens/best_sellers.dart:
```diff
@@ -0,0 +1,83 @@
+import '../../view_model/best_seller_state.dart';
+
+class BestSellers extends StatelessWidget {
+  final BestSellerCubit bestSellerCubit = getIt<BestSellerCubit>();
+
+  BestSellers({super.key});
```
don't make constructor not const this affect on performance use stateful widget instead or make best seller cubit in build method

ahmedelazab1220 commented on lib/features/most_selling_products/presentation/view/screens/best_sellers.dart:
```diff
@@ -0,0 +1,83 @@
+          child: BlocBuilder<BestSellerCubit, BestSellerState>(
+            builder: (context, state) {
+              if (state.baseState is BaseLoadingState) {
+                return const Center(child: CircularProgressIndicator());
+              }
+              if (state.baseState
+                  is BaseSuccessState<List<BestSellerProductEntity>>) {
+                final products =
+                    (state.baseState
+                            as BaseSuccessState<List<BestSellerProductEntity>>)
+                        .data;
+                if (products == null || products.isEmpty) {
+                  return Center(
+                    child: Text(
+                      LocaleKeys.NoProducts.tr(),
+                      style: Theme.of(context).textTheme.bodyMedium,
+                    ),
+                  );
+                }
+                return GridView.builder(
+                  gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
+                    crossAxisCount: 2,
+                    crossAxisSpacing: 17,
+                    mainAxisSpacing: 17,
+                    childAspectRatio: 0.8,
+                  ),
+                  itemCount: products.length,
+                  itemBuilder: (context, index) {
+                    final product = products[index];
+                    return ProductItem(productEntity: product);
+                  },
+                );
+              } else if (state.baseState is BaseErrorState) {
+                AppDialogs.showFailureDialog(
+                  context,
+                  message: (state.baseState as BaseErrorState).errorMessage,
+                );
+              }
+              return const SizedBox.shrink();
+            },
```
make that in seprate file is better you can handle states by other to improve user friendly

ahmedelazab1220 commented on lib/features/most_selling_products/presentation/view/widgets/product_item.dart:
```diff
@@ -0,0 +1,93 @@
+                  child: CachedNetworkImage(imageUrl: productEntity.imgCover!),
+                ),
+              ),
+            ),
```
this is not good way use responsive is better you can use asceptRatio is more better

ahmedelazab1220 commented on lib/features/most_selling_products/presentation/view_model/best_seller_cubit.dart:
```diff
@@ -0,0 +1,39 @@
+    : super(BestSellerState(baseState: BaseInitialState()));
+
+  void doIntent(BestSellerAction action) {
+    if (action is BestSellerRequestAction) {
+      _getBestSellers();
+    }
```
use switch cases is better

ahmedelazab1220 commented:
Great work Eslam 👏 
Keep goind 💯

EslamMagdy12 committed:
refactor(best_sellers): restructure best sellers screen and improve responsiveness

EslamMagdy12 committed:
refactor(most-selling): Improve product item UI and adjust grid layout

- Enhance product item widget by adding placeholder and error handling for images using CachedNetworkImage.
- Adjust grid layout in best sellers body to improve spacing and aspect ratio of product items.

EslamMagdy12 committed:
refactor(most-selling): removed unused imports.

Commit:
Merge branch 'developement' into feature/fa-11-most-selling-products

Commit:
refactor(products): reorganize product item component and update best seller handling

ahmedelazab1220 reviewed (approved):
Great work ya Ammar & Eslam 👏 👏 
keep going

Commit:
refactor(best_sellers): enhance best seller entity conversion and improve UI handling for product display

Commit:
refactor : localization keys for improved readability and consistency

- Cleaned up whitespace in locale_keys.g.dart for better formatting.
- Adjusted line breaks for long constant strings to enhance readability.

Refactor Retrofit client constructors for consistency

- Simplified constructor definitions in auth_retrofit_client.g.dart, categories_retrofit_client.g.dart, home_retrofit_client.g.dart, best_seller_retrofit_client.g.dart, and occasion_retrofit_client.g.dart by removing unnecessary line breaks.

Refactor JSON serialization/deserialization for clarity

- Improved formatting in generated JSON serialization files (categories_response.g.dart, category.g.dart, product.g.dart, products_response.g.dart, home_response_dto.g.dart, product_dto.g.dart, best_seller_products.g.dart, occasions_response_dto.g.dart, products_response_dto.g.dart) for better readability.
- Ensured consistent indentation and line breaks for better maintainability.

Update categories repository interface for clarity

- Reformatted method signatures in categories_repo.dart for improved readability.
